### PR TITLE
Affine Transform Refactoring

### PR DESCRIPTION
**Pull request by mibaumgartner**
## Short Description
This PR performs a refactoring of the affine transformations.
In the following X refers to 'scale', 'rotation', 'translation'

Proposed steps:
1. rename `_format_X` in `utils.affine` to `_create_X` and move them to `transforms.function.affine`
2. refactor _format_X/_create_X a little bit for easier understanding ...
    - this should fix a bug where it is not possible to define the rotation per axis
    - limit input to one of the following (no complete rotation matrices allowed): single scalar, scalar per 
       dim, scalar per batch element, scalar per batch element per dim
3. Double check all doc strings and make them more precise
    - translation needs a unit: see later
   - rotation needs additional documentation in case of 3D where multiple dims are specified ... and should be fixed fro scale param to rotation param

4. double check if scale needs to be inverted
5. translation:  I would like to pass in the number of pixels to shift (right now it only supports a float which shifts the image by that percentage -> shift depend on image size)
6. How to perform rotation around multiple axis (behaviour should replicate bg for compatibility and we may add additional options, e.g. different rotation orders/format, rorate around a unit vector ... )
7. additional unit tests where the behaviour is tested for some special cases: test intensity values not only shape

What do you think about this @justusschock ?

### PR Implementer
This is a small checklist for the implementation details of this PR.
If you submit a PR, please look at these points (don't worry about the `RisingTeam`
and `Reviewer` workflows, the only purpose of those is to have a compact view of
the steps)

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/PhoenixDL/rising/blob/master/CONTRIBUTING.md).

- [x] Implementation
    - [x] 1.
    - [x] 2.
    - [x] 3.
    - [x] 4.
    - [x] 5.
    - [x] 6.
    - [x] 7.
- [x] Docstrings & Typing
- [x] Check `__all__` sections and `__init__`
- [x] Unittests (look at the line coverage of your tests, the goal is 100%!)
- [x] Update notebooks & documentation if necessary
- [x] Pass all tests
- [ ] Add the checksum of the last implementation commit to the Changelog

### RisingTeam
<details>
  <summary>RisingTeam workflow</summary>

  - [x] Add pull request to project (optionally delete corresponding project note)
  - [x] Assign correct label (if you don't have permission to do this, someone will do it for you. 
      Please make sure to communicate the current status of the pr.)
  - [x] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)
</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>
  
  - [x] Do all tests pass? (Unittests, NotebookTests, Documentation)
  - [x] Does the implementation follow `rising` design conventions?
  - [x] Are the tests useful? (!!!) Are additional tests needed? 
        Can you think of critical points which should be covered in an additional test?
  - [x] Optional: Check coverage locally / Check tests locally if GPU is necessary to execute
</details>

### CODE REVIEW

**Comment by justusschock** [comment]
I'm okay with 1.
In 2. it should be possible to specify an angle per axis in the current implementation if you pass it as a list. I'm okay with the other points.
3. And 4. Are also fine.
In 5. I'm not sure if this is really true. If it is, this is definitely a limitation introduced by PyTorch and wen can only encounter this if we manually compute the float each time. This requires more matrix computations on the other hand, even if the spatial size is same for all batches.
6. And 7. Are fine again

**Comment by mibaumgartner** [comment]
2. `trafo = Rotate([0, 45, 0], degree=True, adjust_size=False)` raises a ValueError from the assert statement.  (also tried a tensor and degree=False) [btw. this problem also occurs with translation and scaling, but maybe its only in 3D because the tests pass for the 2D case (Scaling is tested there with multiple axis)]
5. I did some testing in the notebook and the resulting values are matching my hypothesis (event though I think this is not mentioned inside the documentation). The computational burden should be negligible, its just a normalisation of values ... should be doable by overwriting `assemble_matrix` of the translation where the super class is called and the resulting matrix is normalised to the image shape

File: rising/transforms/functional/affine.py [hunk 1]
```diff
@@ -120,8 +116,7 @@ def affine_image_transform(image_batch: torch.Tensor,
         missing_dims = len(image_batch.shape) - len(image_size)
         new_size = (*image_batch.shape[:missing_dims], *new_size)
 
-    matrix_batch = matrix_batch.to(device=image_batch.device,
-                                   dtype=image_batch.dtype)
+    matrix_batch = matrix_batch.to(image_batch)
```
Comment (justusschock) [inline rising/transforms/functional/affine.py]: From which PyTorch version on is this supported? I want to avoid having some implicit constraints here

File: rising/utils/affine.py [hunk 2]
```diff
@@ -106,28 +109,6 @@ def points_to_cartesian(batch: torch.Tensor) -> torch.Tensor:
     return batch[..., :-1] / batch[..., -1, None]
 
 
-def matrix_revert_coordinate_order(batch: torch.Tensor) -> torch.Tensor:
```
Comment (justusschock) [inline rising/utils/affine.py]: We will still need this, if we want to apply it to points.

Because the PyTorch grid sample and affine grid expects the matrix to be in xy(z) and the tensor to transform in BxCx(Dx)HxW. If we now want to apply it to points, we should also use the (z)yx notation to be consistent with the images. We can therefore choose, to internally convert the points (and convert them back afterwards) or to convert the matrix with this (which is, what I actually prefer)

File: rising/transforms/affine.py [hunk 3]
```diff
@@ -572,16 +574,105 @@ def assemble_matrix(self, **data) -> torch.Tensor:
...
                 whole_trafo = matrix
             else:
                 whole_trafo = torch.bmm(whole_trafo, matrix)
-
         return matrix_to_cartesian(whole_trafo)
 
-# TODO: Add transforms around image center
-# TODO: Add Resize Transform
+
+class Resize(Scale):
```
Comment (justusschock) [inline rising/transforms/affine.py]: Probably I won't insert this trafo in his PR, but in #29 , where we will sort out everything else and keep this as a refactoring only-PR

File: rising/transforms/functional/affine.py (hunk 1 above)
Comment (mibaumgartner) [reply rising/transforms/functional/affine.py]: Supported since the beginning / since `to` is supported (0.4.0)

File: rising/utils/affine.py (hunk 2 above)
Comment (mibaumgartner) [reply rising/utils/affine.py]: Will be handled as discussed in our telco :)

File: rising/transforms/affine.py [hunk 4]
```diff
@@ -37,55 +36,15 @@ def __init__(self, scale: AffineParamType = None,
...
-                * a single parameter (as float or int), which will be
-                    replicated for all dimensions and batch samples
-                * a single parameter per sample (as a 1d tensor), which will
-                    be replicated for all dimensions
-                * a single parameter per dimension (either as 1d tensor or as
-                    2d transformation matrix), which will be replicated for
-                    all batch samples
-            None will be treated as a translation offset of 0
         matrix : torch.Tensor, optional
             if given, overwrites the parameters for :param:`scale`,
```
Comment (justusschock) [inline rising/transforms/affine.py]: There are no parameters for scale, rotation and translation here. Either they are missing or (more likely) the doctoring is not up to date

File: rising/transforms/affine.py [hunk 5]
```diff
@@ -252,7 +210,226 @@ def __radd__(self, other):
...
+        for trafo in self.transforms:
+            matrix = matrix_to_homogeneous(trafo.assemble_matrix(**data))
+            if whole_trafo is None:
+                whole_trafo = matrix
+            else:
+                whole_trafo = torch.bmm(whole_trafo, matrix)
+        return matrix_to_cartesian(whole_trafo)
+
+
+class BaseAffine(Affine):
```
Comment (justusschock) [inline rising/transforms/affine.py]: Why did you decide to outsource this to a separate class? IMO a complete class may be overkill for just this one function. Couldn't just include it in the main `Affine`?

File: rising/transforms/functional/affine.py [hunk 6]
```diff
@@ -1,15 +1,411 @@
...
+        sin value to use for rotation matrix, [1]
+    cos: Tensor
+        cos value to use for rotation matrix, [1]
+
+    Returns
+    -------
+    Tensor:
+        rotation matrix, [3, 3]
+    """
+    return torch.tensor([[1., 0., 0.],
```
Comment (justusschock) [inline rising/transforms/functional/affine.py]: can we please provide dtype and device here? otherwise they may be created on CPU and with inappropriate types and we have to transfer them later, which is inefficient

File: rising/transforms/functional/affine.py [hunk 7]
```diff
@@ -1,15 +1,411 @@
...
+        sin value to use for rotation matrix, [1]
+    cos: Tensor
+        cos value to use for rotation matrix, [1]
+
+    Returns
+    -------
+    Tensor:
+        rotation matrix, [3, 3]
+    """
+    return torch.tensor([[cos.clone(), 0., sin.clone()],
```
Comment (justusschock) [inline rising/transforms/functional/affine.py]: dtype and device?

File: rising/transforms/functional/affine.py [hunk 8]
```diff
@@ -1,15 +1,411 @@
...
+                         [0., 1., 0.],
+                         [-sin.clone(), 0., cos.clone()]])
+
+
+def create_rotation_3d_2(sin: Tensor, cos: Tensor) -> Tensor:
```
Comment (justusschock) [inline rising/transforms/functional/affine.py]: dtype and device?

File: rising/transforms/functional/affine.py [hunk 9]
```diff
@@ -1,15 +1,411 @@
...
+        sin value to use for rotation matrix, [1]
+    cos: Tensor
+        cos value to use for rotation matrix, [1]
+
+    Returns
+    -------
+    Tensor
+        rotation matrix, [2, 2]
+    """
+    return torch.tensor([[cos.clone(), -sin.clone()], [sin.clone(), cos.clone()]])
```
Comment (justusschock) [inline rising/transforms/functional/affine.py]: dtype and device?

File: rising/utils/affine.py [hunk 10]
```diff
@@ -106,28 +104,6 @@ def points_to_cartesian(batch: torch.Tensor) -> torch.Tensor:
...
```
Comment (justusschock) [inline rising/utils/affine.py]: we will need this, when we revert the order by a flag (as we discussed) to optionally make it consistent with PyTorch

File: .github/workflows/notebook_tests.yml [hunk 11]
```diff
@@ -9,7 +9,8 @@ jobs:
       max-parallel: 4
       matrix:
         python-version: [3.7]
-
+    env:
```
Comment (justusschock) [inline .github/workflows/notebook_tests.yml]: What does this do?

File: rising/transforms/affine.py [hunk 12]
```diff
@@ -17,13 +19,10 @@
 
 
 class Affine(BaseTransform):
-    def __init__(self, scale: AffineParamType = None,
-                 rotation: AffineParamType = None,
-                 translation: AffineParamType = None,
-                 matrix: torch.Tensor = None,
+    def __init__(self,
```
Comment (justusschock) [inline rising/transforms/affine.py]: What about the flag we discussed to switch between the 'naive' way (where you provide things for each dim in order) and the Pytorch way, where you provide it in xyz order independent of the actual axes order in your tensor?

File: rising/transforms/affine.py (hunk 4 above)
Comment (mibaumgartner) [reply rising/transforms/affine.py]: See below

**Comment by justusschock** [comment]
Nother thing that came to my mind, but is not essential: We should include description text in our notebooks :)

File: rising/transforms/affine.py (hunk 5 above)
Comment (mibaumgartner) [reply rising/transforms/affine.py]: I like the single responsibility principle :) Furthermore, this will help us in the future. With this outsourcing it is possible for us to add new representations in the future without rewriting the base class. The BaseAffine is just one special case how one can visualise transformations, e.g. rotation can also be modelled by an rotation around a parametrised axis (rotate 50 deg around [1,1,1])

File: .github/workflows/notebook_tests.yml (hunk 11 above)
Comment (mibaumgartner) [reply .github/workflows/notebook_tests.yml]: This just adds an environment flag for the notebook which disables the visualisation during testing

File: rising/transforms/affine.py (hunk 12 above)
Comment (mibaumgartner) [reply rising/transforms/affine.py]: will add that with the next commit

**Comment by mibaumgartner** [comment]
Currently, the notebook is more like a visual test of the transformations. :) When the GridTransforms are ready, I will refactor it to be pretty :D
